File: kuyhochung/1. 매주 문제풀이/52주차/괄호회전하기.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

bool is_correct_string(string input) {
    stack<char> buffer;
    for (int i = 0; i < input.size(); i++) {
        if ((input[i] == '(') || (input[i] == '{') || (input[i] == '['))
            buffer.push(input[i]);
        else {
            while(1) {
                if (buffer.empty()) return false;
                char temp = buffer.top();
                buffer.pop();
                if (input[i] == ')') {
                    if (temp == '(') break;
                }
                if (input[i] == '}') {
                    if (temp == '{') break;
                }
                if (input[i] == ']') {
                    if (temp == '[') break;
                }
            }
        }
    }
    if (buffer.empty()) return true;
    else return false;
}

int solution(string s) {
    int answer = 0;
    
    if (s == "") return 0;
    
    for (int i = 0; i < s.size(); i++) {
        if (is_correct_string(s)) answer++;
        string temp = "";
        temp = s.substr(1);
        temp += s[0];
        s = temp;
    }
    
    return answer;
}

File: JongHo/Programmers/괄호회전하기.py
def check(g):
    stack = []
    for i in range(len(g)-1, -1, -1):
        if not stack:
            stack.append(g[i])
            continue
            
        if g[i] + stack[-1] == "()":
            stack.pop()
        elif g[i] + stack[-1] == "{}":
            stack.pop()
        elif g[i] + stack[-1] == "[]":
            stack.pop()
        else:
            stack.append(g[i])
            
    if stack:
        return False
    return True
                
        

def solution(s):
    s = list(s)
    answer = 0
    for _ in range(len(s)):
        if check(s):
            answer += 1
        span = s.pop(0)
        s.append(span)
    
    return answer